### PR TITLE
make monolith ids constant

### LINKED ISSUE #1251: make monolith ids constant from the perspective of the balancers
Currently, monoliths are assigned a unique `MonolithId` by balancers. However, this makes it extremely difficult to correlate which monolith is which just by querying the balancers. 

- Change the monolith so that it generates it's own id, and conveys it to the balancer in the init message.
- Change the balancer to use this id instead of generating it.
- Add unit tests to the monolith to assert for this new behavior.
    - useful related docs: https://jestjs.io/docs/mock-functions
    - see tests in `clientmanager.spec.ts`

Also update the spec doc to reflect these changes.

### PR DESCRIPTION
**Pull request by Victor-M-Giraldo**
closes #1251

### CODE REVIEW

**Comment by dyc3** [comment]
Yeah that sounds right

**Comment by Victor-M-Giraldo** [comment]
How is the id defined in the constructor for `BalancerConnection` any different than a `MonolithID`?

Also, is there an example somewhere in the code of reading the payload, or how would I access that id to use in `connection.rs` rather than generating it in there?

File: server/balancer.ts [hunk 1]
```diff
@@ -76,6 +76,7 @@ class BalancerManager {
 			payload: {
 				port: conf.get("port"),
 				region: conf.get("balancing.region"),
+				id: uuidv4(),
```
Comment (dyc3) [inline server/balancer.ts:79]: This still makes the id random for every balancer it connects to. It needs to be constant across reconnects. It's ok if the id is not constant across service restarts (eg don't store it in a file, it being a global is fine).

File: crates/ott-balancer/src/connection.rs [hunk 2]
```diff
@@ -126,18 +123,21 @@ async fn connect_and_maintain(
 
         // Handle connection initialization
         let mut outbound_rx;
+        let id_clone;
```
Comment (dyc3) [inline crates/ott-balancer/src/connection.rs:126]: nit: 
```suggestion
        let monolith_id;
```

**Comment by Victor-M-Giraldo** [comment]
Having some trouble writing the test for this. Might ask for help tomorrow.

**Comment by dyc3** [comment]
The most valuable test here would probably be a unit test in `clientmanager.spec.ts` that creates 2 `BalancerConnectionMock` and see if they both get sent the same id.

**Comment by Victor-M-Giraldo** [comment]
How can I run only tests for a certain file? Is that possible?

**Comment by dyc3** [comment]
Yeah should be possible. Look up the docs for `jest` or `cargo test`, depending if you're working with ts or rust, respectively. I can never get it to work reliably with jest

**Comment by Victor-M-Giraldo** [comment]
Sounds good. I'll take a look. Just want to make sure the test is working without running the entire suite.

Also, the monolith IDS should be different, correct?

**Comment by dyc3** [comment]
Monolith ids should be the same within the context of the monolith. Across multiple monoliths, their ids should be different.

**Comment by Victor-M-Giraldo** [comment]
Okay, that's what I thought.

**Comment by Victor-M-Giraldo** [comment]
Figure this probably isn't what you mean. 

Is it something to do with `sendMock`?

File: server/tests/unit/clientmanager.spec.ts [hunk 3]
```diff
@@ -172,3 +172,15 @@ describe("BalancerManager", () => {
 		expect(balancerManager.balancerConnections).not.toContain(con2);
 	});
 });
+
+describe("MonolithId", () => {
+	it("should have unique IDS for each instance of a monolith", () => {
```
Comment (dyc3) [inline server/tests/unit/clientmanager.spec.ts:177]: From this context, there is only 1 monolith: the one that is running. The client manager should send the same id to both of these mock connections.

**Comment by Victor-M-Giraldo** [comment]
I only see stuff about the client ID there. Or is that what you meant?

**Comment by dyc3** [comment]
No, I'm not talking about the client id, I'm talking about the goal of this pr.

**Comment by Victor-M-Giraldo** [comment]
Alright.

**Comment by Victor-M-Giraldo** [comment]
Think that works.

File: server/tests/unit/clientmanager.spec.ts [hunk 4]
```diff
@@ -181,10 +181,11 @@ describe("MonolithId", () => {
 		balancerManager.addBalancerConnection(mock1);
 		balancerManager.addBalancerConnection(mock2);
 
-		const id1 = mock1.sendMock.mock.calls[0][0].payload.id;
-		const id2 = mock2.sendMock.mock.calls[0][0].payload.id;
+		const id1 = mock1.sendMock.mock.calls[0][0].filter((msg: MsgM2B) => msg.type == "init")[0]
+			.payload.id;
+		const id2 = mock2.sendMock.mock.calls[0][0].filter((msg: MsgM2B) => msg.type == "init")[0]
+			.payload.id;
```
Comment (dyc3) [inline server/tests/unit/clientmanager.spec.ts:187]: this is filtering the wrong one. you need to filter the calls, not the call arguments

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: server/tests/unit/clientmanager.spec.ts (hunk 4 above)
Comment (dyc3) [reply server/tests/unit/clientmanager.spec.ts:187]: It's ok to leave it the way it was

**Comment by dyc3** [comment]
if you want to rebase on master, I just pushed a commit that adds type annotations to `sendMock`

File: server/tests/unit/clientmanager.spec.ts [hunk 5]
```diff
@@ -201,3 +202,18 @@ describe("BalancerManager", () => {
...
+		const mock2 = new BalancerConnectionMock();
+
+		balancerManager.addBalancerConnection(mock1);
+		balancerManager.addBalancerConnection(mock2);
+
+		const id1 = (mock1.sendMock.mock.calls[0][0].payload as M2BInit).id;
+		const id2 = (mock2.sendMock.mock.calls[0][0].payload as M2BInit).id;
+
+		expect(id1).toEqual(id2);
+	});
```
Comment (dyc3) [inline server/tests/unit/clientmanager.spec.ts:218]: this should also assert that id1 and id2 are both not undefined

**Review by dyc3 — CHANGES_REQUESTED** [review]

**Review by dyc3 — APPROVED** [review]

**Pull request merged**